### ISSUE TITLE
quantized_llama.rs uses wrong RoPE convention for NEOX-style models (Qwen2, Falcon, Phi, etc.)

### ISSUE THREAD

**Issue by joelteply** [huggingface/candle#3410]
## Description

Candle's GGUF quantized inference produces significantly degraded output compared to llama.cpp for the same model file and prompt. The model generates ~30 tokens of coherent text then degenerates into repetition, while llama.cpp produces perfect, complete output.

## Environment

- Candle version: 0.9.2
- Model: Qwen2.5-Coder-14B (head-pruned compacted variant, Q5_K_S)
- Tested on: Metal (macOS M-series) and CPU — both produce identical degraded output
- llama.cpp: latest main branch, CUDA (RTX 5090)

## Reproduction

**Same GGUF file, same prompt, same temperature (0.1), same repeat_penalty (1.3).**

### Candle output (Metal or CPU):
```
code/write: {path: "app.py",assistant
\```python
from flask import Flask, jsonify
app = Flask(__name: "app.py
\```python
from flask
from flask
```
Starts correct, degenerates into repetition after ~30 tokens.

### llama.cpp output (CUDA):
```xml
<tool_use>
<tool_name>code/write</tool_name>
<parameters>
{"filePath": "hello.py", "content": "from flask import Flask\napp = Flask(__name__)\n\n@app.route('/')\ndef hello():\n    return 'Hello, World!'\nif __name__ == '__main__':\n    app.run(debug=True)"}
</parameters>
</tool_use>
```
Perfect XML tool call, correct code, clean EOS stop.

## Analysis

The model metadata is correct (verified via gguf-py):
- `attention.head_count = 25` (pruned from 40)
- `attention.head_count_kv = 5` (pruned from 8)
- `attention.key_length = 128`
- `attention.value_length = 128`

Both Candle CPU and Metal produce identical degraded output, ruling out Metal-specific bugs. The divergence is in Candle's core quantized inference path vs llama.cpp's.

Possible causes:
1. **Q5_K_S dequantization numerical differences** accumulating through 48 layers
2. **RoPE implementation divergence** (frequency computation or application)
3. **Attention computation differences** (mask handling, softmax precision)
4. **Sampling path differences** (logits processing, temperature application)

## Model

The GGUF is published at [continuum-ai/qwen2.5-coder-14b-compacted](https://huggingface.co/continuum-ai/qwen2.5-coder-14b-compacted) (Q5_K_S, 8.9GB). Note: this model has non-standard attention dimensions (25 Q-heads instead of 40) due to head pruning. llama.cpp required a [one-line patch](https://github.com/huggingface/candle/issues/XXXX) to accept the tensor shapes, but produces correct output once loaded.

**Comment by joelteply** [comment]
## Investigation Update — Root Cause Analysis

We've done extensive A/B testing with the same GGUF on llama.cpp (CUDA, RTX 5090) vs Candle 0.9.2 (Metal + CPU, Apple Silicon). Key findings:

### Confirmed: Tokenization is identical
Both implementations produce exactly 187 tokens with the same IDs. The prompt input is byte-for-byte identical.

### Confirmed: RoPE frequencies match to 15 digits
Candle's independent `powf()` and llama.cpp's iterative `theta *= theta_scale` produce frequencies that agree to ~3.6e-15 relative error. Not a factor.

### Confirmed: Q5_K dequantization logic is identical
Same block parsing, same bit extraction, same scaling formula. The NEON SIMD path matches llama.cpp's NEON path structurally.

### Root Cause 1: RMSNorm accumulates in f32 (should be f64)
llama.cpp accumulates sum-of-squares in `ggml_float` (double/f64), then converts to f32.
Candle accumulates entirely in f32 (`candle_nn::ops::rms_norm` → `.sum::<f32>()`).

For hidden_size=5120, this introduces ~7.7e-6 relative error per layer. Across 48 layers, this compounds significantly. We patched this in our vendored RmsNorm and confirmed it **changes the output** (different tokens selected), proving this is a real precision source.

### Root Cause 2: Repetition penalty applied on probabilities (should be logits)
Candle's `LogitsProcessor::sample_f()` callback operates on **probabilities** (after softmax). llama.cpp applies repetition penalty on **logits** (before softmax).

With temperature 0.1, softmax makes the distribution very peaked:
- On logits: `20.0 / 1.3 = 15.4` → drops below competitors, softmax redistributes effectively
- On probabilities: `0.95 / 1.3 = 0.73` → still dominant after renormalization, penalty is ineffective

We patched this and confirmed it fixes specific token divergences (e.g., `Flask(__name__)` vs `Flask(__name:`).

### Remaining Issue
Even with both fixes, the model still degenerates after ~30 tokens. The cumulative f32 precision differences across 48 layers in all operations (matmul accumulation, softmax, SiLU activation) shift the logit distribution enough that the model goes off its coherent generation trajectory.

The model is proven correct — llama.cpp produces perfect, complete tool-call output with clean EOS on the same GGUF.

### Reproduction Setup
- Model: [continuum-ai/qwen2.5-coder-14b-compacted](https://huggingface.co/continuum-ai/qwen2.5-coder-14b-compacted) (Q5_K_S, 25 Q-heads, 5 KV-heads, head_dim=128)
- llama.cpp: latest main, patched for non-standard head dims
- Candle: 0.9.2
- Prompt: 187 tokens (Qwen2 ChatML template with tool instructions)

**Comment by joelteply** [comment]
## Definitive Comparison: Same Mac, Same Metal GPU, Same GGUF

We built a patched llama.cpp locally (to handle compacted head dims) and ran both on the **exact same hardware** (M1 Pro, Metal):

### Top-10 logits after 187-token prefill:

**Candle 0.9.2:**
```
1. endoftext[151643]  18.788    ← control token dominates!
2. "code"[1851]       15.291
3. "assistant"[77091] 14.857
4. im_start[151644]   14.552
8. "<"[27]            13.676
10."I"[40]            12.905
EOS[151645]           10.326
```

**llama.cpp (Metal, same Mac):**
```
1. "I"[40]            19.866    ← natural language token
2. "Let"[10061]       15.347
3. "<"[27]            13.220
4. "Sure"[39814]      12.482
endoftext[151643]      7.073
EOS[151645]            4.481
```

### Key divergences:
| Token | Candle | llama.cpp | Δ |
|-------|--------|-----------|---|
| endoftext (151643) | 18.788 | 7.073 | **+11.7** |
| "I" (40) | 12.905 | 19.866 | **-7.0** |
| EOS (151645) | 10.326 | 4.481 | **+5.8** |

### Layer-by-layer binary search (Candle):
- Layers 0-46: endoftext NOT in top 10 ✓
- Layer 47: endoftext appears at rank 4 (logit 11.2)
- Layer 48: endoftext JUMPS to rank 1 (logit 18.8)

This is a **systematic directional bias** — Candle inflates special/control token logits while deflating natural language tokens. The bias compounds with layer count, explaining why it only manifests in deep models (48 layers).

### What we've verified matches:
- ✅ Tokenization: identical (187 tokens, same IDs)
- ✅ RoPE frequencies: match to 15 digits
- ✅ Q5_K dequantization logic: identical
- ✅ Embedding values: manually verified, match exactly
- ✅ Model architecture: shapes, head dims, bias loading all correct
- ✅ Tensor types identical between layers 46 and 47

### What's different:
The logit distributions are fundamentally incompatible — not close-but-reordered, but completely different rankings. This points to a **structural bug** in Candle's quantized forward pass that creates a per-layer directional bias toward special tokens.

**Comment by joelteply** [comment]
This is a much deeper issue in candle accuracy. Digging into it now. Finding a lot of hacks in this thing, which compound through exponentiation of softmax and other layers. It should be 1:1 with llama.cpp. I've created a fork and will figure this out. Quantization just degrades and I suspect a lot of the other issues presented are from the same root causes.

This project has been a godsend for local inference, LoRA and other refinements, but we must fix these, because quantization strategies are the best method of running these models on lower end macbooks.

**Comment by joelteply** [comment]
## Root Cause Found: Wrong RoPE Convention for NEOX-style Models

After extensive binary-search debugging (dumping intermediate tensors at every layer step and comparing with llama.cpp), we found the root cause.

### The Bug

`quantized_llama.rs` always uses `rope_i` (interleaved RoPE), which pairs dimensions `(2i, 2i+1)`. This is correct for standard Llama models (`rope_type=0`, NORM), but **wrong for Qwen2** and other NEOX-style architectures (`rope_type=2`), which pair `(i, i+d/2)`.

In llama.cpp, the rope type is determined per-architecture (see `llama_model_rope_type()` in `llama-model.cpp`):
- **NORM (interleaved)**: Llama, Mistral, DeepSeek, etc.
- **NEOX (non-interleaved)**: Qwen, Qwen2, Falcon, Phi, StableLM, BERT variants, OLMo2, etc.

Candle's `quantized_llama.rs` ignores this distinction and always uses interleaved.

### The Fix

Read the model architecture from GGUF metadata (`general.architecture`) and dispatch:

```rust
fn apply_rotary_emb(&self, x: &Tensor, index_pos: usize) -> Result<Tensor> {
    let cos = self.cos.narrow(0, index_pos, seq_len)?;
    let sin = self.sin.narrow(0, index_pos, seq_len)?;
    if self.rope_is_neox {
        // GPT-NeoX: pairs (i, i+d/2). Qwen2, Falcon, Phi, etc.
        candle_nn::rotary_emb::rope(&x.contiguous()?, &cos, &sin)
    } else {
        // Interleaved: pairs (2i, 2i+1). Standard Llama.
        candle_nn::rotary_emb::rope_i(&x.contiguous()?, &cos, &sin)
    }
}
```

Where `rope_is_neox` is set during model loading based on the architecture string:
```rust
let rope_is_neox = matches!(arch.as_str(),
    "qwen" | "qwen2" | "qwen2moe" | "qwen3" | "qwen3moe" |
    "falcon" | "phi" | "phi2" | "phi3" | "stablelm" | ...
);
```

### Verification

After the fix, Candle matches llama.cpp to **<0.01 logit precision** across all 152K vocab tokens after 48 layers:

| Token | Candle (fixed) | llama.cpp | Δ |
|-------|--------|-----------|---|
| "I" (40) | 19.868 | 19.866 | **0.002** |
| "Let" (10061) | 15.340 | 15.347 | 0.007 |
| endoftext (151643) | 7.074 | 7.073 | **0.001** |
| EOS (151645) | 4.481 | 4.481 | **0.000** |

Top-10 rankings are now **identical**. The +11.7 logit inflation on `endoftext` is gone.

### How We Found It

1. Dumped layer 0 intermediate tensors from both Candle and llama.cpp (using eval callback)
2. `attn_norm` (RmsNorm output): **bit-identical** ✓
3. Q/K/V projections + bias: matched to 0.00004 ✓
4. `kqv_out` (attention output before Wo): **diverged** — max_abs=0.53, varying by head
5. The per-head error pattern pointed to RoPE (different frequency bands → different error magnitudes)
6. Confirmed: llama.cpp's Metal `kernel_rope_neox` uses `src[args.n_dims/2]` pairing, Candle's `rope_i` uses adjacent pairing

### Scope

This bug affects **all NEOX-style GGUF models** loaded through `quantized_llama.rs`, not just Qwen2. This includes Falcon, Phi-2/3, StableLM, OLMo2, and many others. The `candle_nn::rotary_emb::rope` function (non-interleaved) already exists and works correctly — it just wasn't being called.

Will prepare a PR once we've done full end-to-end testing.
